### PR TITLE
Fixes #37852 - Support Rails 7.0

### PR DESCRIPTION
**Pull request by ofedoren**
#### What are the changes introduced in this pull request?

We're updating Rails version in Foreman: https://github.com/theforeman/foreman/pull/10299 (please visit it first if you have any questions/suggestions, there are many comments regarding the changes)

This PR tries to make Katello Rails 7 friendly (compatible).

#### Considerations taken when implementing this change?

1. Mostly removing deprecated usage of `#<<` for ActiveModel::Errors objects
2. Just a heads up regarding https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#digest-class-for-activesupport-digest-changing-to-sha256. How it affects this plugin is more or less here: https://github.com/theforeman/foreman/pull/10299#discussion_r1832768551

#### What are the testing steps for this pull request?

Ensuring CI is green.

If you really want some :cookie:, test some workflows live.

### CODE REVIEW

File: test/lib/validators/hostgroup_kickstart_repository_validator_test.rb
Comment (ekohl) [inline test/lib/validators/hostgroup_kickstart_repository_validator_test.rb:66]:
```suggestion
      assert_empty @content_facet.errors.map(&:message)
```

**Review by ekohl — COMMENTED** [review]
I see lots of changes around the error handling and I think that's already Rails 6.1 compatible. Perhaps something that we can already merge to reduce the size of this PR?

**Comment by chris1984** [comment]
@SatelliteQE/TeamPhoenix

**Comment by chris1984** [comment]
@SatelliteQE/team-phoenix

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 9132
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman --team Phoenix-content --external-logging
Test Result: = 6 skipped, 3973 deselected, 5492 warnings, 1490 errors in 4906.50s (1:21:46) =
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 9133
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman --team Phoenix-subscriptions --external-logging
Test Result: ======= 4779 deselected, 5492 warnings, 690 errors in 4923.46s (1:22:03) =======
```

**Comment by ekohl** [comment]
@chris1984 I doubt PRT is going to be of any use. You need updated Foreman and Rails and we haven't packaged that.

**Comment by ofedoren** [comment]
The test failure is unrelated to this PR, it for some reason wasn't caught by GHA for https://github.com/theforeman/foreman_remote_execution/pull/913

File: app/lib/katello/util/data.rb
Comment (ofedoren) [inline app/lib/katello/util/data.rb:9]: This forces SHA1 for hexdigest since it was default in Rails < 7. This makes sure Katello behaves the same way and the tests are green.

I'd ask maintainers to check the needed changes for this plugin to support SHA256. Afterwards this line can be switched back (until Rails upgrades the default to SHA512).

Comment (ekohl) [reply app/lib/katello/util/data.rb:9]: @parthaa said in https://community.theforeman.org/t/rails-7-upgrade/37771/27 that the actual value doesn't matter, as long as it's consistent.

I wonder if the old code was ever used. https://github.com/rails/rails/commit/82822a34217503336d51b7baab82cd18cf71e435 introduced `ActiveSupport::Digest` and was introduced in Rails 5.2 (by a former Foreman developer). Perhaps a PR to replace it al with direct calls to `ActiveSupport::Digest.hexdigest()` is a better approach.

Having said that, I find it interesting that the class itself still defaults to md5: https://github.com/rails/rails/blob/4df235f7c0149ac7be86580d41a74a7785cb9bb9/activesupport/lib/active_support/digest.rb#L9

Comment (ofedoren) [reply app/lib/katello/util/data.rb:9]: So should I just revert the changes back and simply update IDs in VCR tests?

File: katello.gemspec
Comment (ekohl) [inline katello.gemspec:66]: I split this off into https://github.com/Katello/katello/pull/11222.

**Comment by ekohl** [comment]
I see you split off https://github.com/Katello/katello/pull/11226 into a separate PR, which is great. Are you also planning to split of the removal of require statements? I'd think that's also compatible with our current Zeitwerk and that only scopes this PR to the digest changes.

File: app/lib/katello/util/data.rb
Comment (ianballou) [reply app/lib/katello/util/data.rb]: Since the IDs just need to be random numbers, I think it's fine to use the new SHA 256 default. I was going to suggest continuing to use SHA1 like suggested here so the VCR recordings could be avoided, but that's already done, so I think we're all good.

File: .github/workflows/ruby.yml
Comment (ianballou) [inline .github/workflows/ruby.yml:30]: Comment reminder to revert this back to develop before merging.

**Comment by ekohl** [comment]
When I locally rebase this I see there are some changes remaining where the `template` parameter to `respond_for_show` and `respond_for_index` is changed into `full_template`. Why is that change needed?

**Comment by parthaa** [comment]
> When I locally rebase this I see there are some changes remaining where the `template` parameter to `respond_for_show` and `respond_for_index` is changed into `full_template`. Why is that change needed?

hmm seems connected to 
```patch
diff --git a/app/lib/katello/api/v2/rendering.rb b/app/lib/katello/api/v2/rendering.rb
index 653e41a76e..a987e9ebde 100644
--- a/app/lib/katello/api/v2/rendering.rb
+++ b/app/lib/katello/api/v2/rendering.rb
@@ -41,8 +41,9 @@ module Katello
         def respond_with_template(action, resource_name, options = {}, &_block)
           yield if block_given?
           status = options[:status] || 200
+          template = options[:full_template] || "katello/api/v2/#{resource_name}/#{action}"

-          render :template => "katello/api/v2/#{resource_name}/#{action}",
+          render :template => template,
                  :status => status,
                  :locals => options.slice(:object_name, :root_name, :locals),
                  :layout => "katello/api/v2/layouts/#{options[:layout]}"
```


Guessing relative paths dont work with zeitwerk ?

``` patch
-      respond_for_show(:resource => host, :template => '../../../api/v2/hosts/show')
+      respond_for_show(:resource => host, :full_template => 'katello/api/v2/hosts/show')
```

File: app/lib/katello/util/data.rb
Comment (ofedoren) [reply app/lib/katello/util/data.rb]: I just thought if it will break stuff for upgrading users:
 - IDs are generated or rather mapped (for the same input `A` -> the same output `B`)
 - These IDs are stored somewhere
 - Newer hashing class will create for the same input `A` -> the same output, but different `C`
 - The stored `B` can't be referenced by `A`, since hashed version is `C`.
 
 But if mapping goes only before storing and then the app using only stored values (without any further `#hexdigest` calls on `A` to reference stored value) then yeah, it should be safe. We just would need to update tests whenever new default comes from Rails.

**Comment by ofedoren** [comment]
@ekohl, @parthaa, without this change it just stops working on Rails 7. Not sure why, but this change was safe, straightforward and compatible for other places. It just seemed easier fixing like that then digging to find what change in Rails caused that :/

**Review by ianballou — APPROVED** [review]
This is looking fine to me, I tested it a bit with https://github.com/theforeman/foreman/pull/10299.
Just don't forget to change the foreman version back to develop for testing!

**Comment by ofedoren** [comment]
Sorry for pushing after approval. I've just rebased both Foreman PR and this one, just to ensure it's still green, so I can drop the last commit to make it ready for merge.

**Comment by ofedoren** [comment]
PR is green after rebase, thus dropped the last commit.

Screenshot_20241126_142427

**Comment by ianballou** [comment]
It looks like a couple of the VCRs need recording again for some reason.

**Comment by ianballou** [comment]
Unfortunately I think there have been new commits added to master that rerecorded the candlepin VCRs.

**Comment by ofedoren** [comment]
Still green after rebase, thus dropping "test" commit:
Screenshot_20241204_165946

**Comment by ekohl** [comment]
The Foreman core PR has been. Anyone with permissions who can merge this?

**Pull request merged**
